dynamiqs/qarray.py

```diff
@@ -0,0 +1,554 @@
+from __future__ import annotations
+
+import warnings
+from abc import abstractmethod
+
+import equinox as eqx
+import jax.numpy as jnp
+import numpy as np
+import qutip as qt
+from jax import Array
+from jax.scipy.linalg import expm
+from jaxtyping import ArrayLike, ScalarLike
+
+from .utils.utils.general import (
+    cosm,
+    dag,
+    isbra,
+    isdm,
+    isket,
+    isop,
+    norm,
+    proj,
+    ptrace,
+    sinm,
+    tensor,
+    tobra,
+    todm,
+    toket,
+    unit,
+)
+
+__all__ = ['QArray', 'DenseQArray']
+
+
+# a decorator that takes a class method f and returns g(f(x))
+def pack_dims(method: callable) -> callable:
+    """Decorator to return a new QArray with the same dimensions as the original one."""
+
+    def wrapper(self: QArray, *args, **kwargs) -> callable:
+        return self.__class__(method(self, *args, **kwargs), dims=self.dims)
+
+    return wrapper
+
+
+class QArray(eqx.Module):
+    r"""Quantum array object. DenseQArray is a wrapper around JAX arrays. It offers
+    convenience methods to handle more easily quantum states
+    (bras, kets and density matrices) and quantum operators.
+    If you come from QuTiP, this is the equivalent of the Qobj class.
+    """
+
+    dims: tuple[int, ...]
+
+    @property
+    @abstractmethod
+    def dtype(self) -> jnp.dtype:
+        """Data type of the quantum state.
```

Comment:
I'm not sure we want to fix documentation in this PR, but in the rest of the library, single-liners start with active verbs (e.g. "Returns the data type of the quantum state.").

Reply:
I don't really like this style but I'll follow it for now and we can discuss it later